syntax = "proto3";
import "nanopb.proto";

message AddRequest {
  uint32 a = 1;
  uint32 b = 2;
}

message AddResponse {
  uint32 result = 1;
}

message Empty{

}

message bankinfo {
    uint32 active = 1;
    uint32 major = 2;
    uint32 minor = 3;
    uint32 patch = 4;
    uint32 crc   = 5;
}

message sw_info{
    bankinfo bank1  = 1;
    bankinfo bank2  = 2;  
}
enum UPD_CMD{
  FLASH_ERASE = 0;
  DATA_CHUNK = 1;
  ABORT_UPGRADE = 2;
  FW_UPGRADE_DONE = 3;
  RESET_DEVICE = 4;
}


enum STATUS{
  NO_ERROR = 0;
  FLASH_ERASE_SUCCESS = 1;
  FLASH_ERASE_FAILED = 2;
  ACK_FOR_CHUNKS = 3;
  FW_CRC_FAILURE = 4;
  FW_CRC_SUCCESS = 5;
  UNEXPECTED_COMMAND =6;
}

message fw_upgrade_req{
  UPD_CMD cmd = 1;
  uint32   fw_size = 2; // Sent when cmd is to FLASH_ERASE
  repeated uint32 data = 3 [(nanopb).max_count = 8];
}


message fw_upgrade_resp{
  UPD_CMD last_cmd = 1;
  STATUS status = 2;
}

enum LED_STATE{
  LED_ON = 0;
  LED_OFF = 1;
  LED_TOGGLE = 2;
}
message led_ctrl_req{
  LED_STATE operation = 1;
}


message proto_msg {
  oneof payload{
    Empty reset_req = 1;
    AddRequest add_req = 2;
    AddResponse add_resp = 3;
    Empty sw_info_req = 4;
    sw_info sw_info_resp = 5;
    fw_upgrade_req upd_req = 6;
    fw_upgrade_resp upd_resp = 7;
    led_ctrl_req led_req = 8; 
  }
}
